### DISCUSSION TITLE
The Case for Spark

### DISCUSSION THREAD

**Discussion by bencroker** [putyourlightson/craft-spark#2] [Announcements]
[**Sprig**](https://putyourlightson.com/sprig) has been a labour of love for the past 4 years, and while it has proven extremely popular, I’ve noticed some shortcomings to its overall approach.

### 1. Its API is coupled to that of htmx.

Sprig builds on top of htmx and is closely coupled to it. As new features (and scope creep) are added to htmx, they are naturally added to Sprig. So while Sprig adds its own concept of Twig components, the mechanism by which it works is heavily dependent on that of htmx.

### 2. It only solves part of the problem.

Sprig provides the ability to re-render Twig templates on the back-end and swap them into the DOM on the front-end, but it lacks pure front-end reactivity. I’ve [written about the htmx API](https://putyourlightson.com/articles/htmx-has-a-javascript-api-btw) and how it can be used to react to events on the page, but most developers still reach for a JavaScript library like [Alpine JS](https://alpinejs.dev/) to facilitate front-end reactivity.

### 3. It uses an antiquated, suboptimal approach.

Sprig adds reactive components to Twig. While the concept is relatively straightforward to grasp, it results in more Twig code rendered than necessary, more HTML sent over the wire than necessary, and more of the DOM swapped out than necessary. Also, swapping multiple parts of the DOM requires either multiple request-response cycles, or the use of out-of-band swaps (a clumsy API that Sprig attempts to simplify but which can still be confusing).

---

[**Spark**](https://github.com/putyourlightson/craft-spark) aims to address the shortcomings above by:

1. Decoupling the mechanism (as much as possible).
2. Solving both back-end _and_ front-end reactivity in a cohesive manner.
3. Using a more modern, optimal approach.

Part of how it does this is by leveraging [Datastar](https://data-star.dev), a JavaScript library that combines the core functionality of [Alpine JS](https://alpinejs.dev/) with that of [htmx](https://htmx.org/). Datastar takes a hypermedia-first approach, meaning that you won’t find history support, JavaScript execution in responses, nor any other “bells and whistles”. This is intentional. By embracing the simplicity of hypermedia, the encapsulation of web components (natively or using [Lit](https://lit.dev/)) and the optimised DOM operations of web browsers, you can build highly performant, hypermedia-driven web apps, without requiring a full-blown JavaScript framework.

Read [Datastar’s getting started guide](https://data-star.dev/guide/getting_started).

Spark uses Datastar for interacting with the back-end (via its [back-end plugins](https://data-star.dev/reference/plugins_backend)), and provides a simple API for which to do so. Take a look at the [usage docs](https://putyourlightson.com/plugins/spark#usage) to see what it looks like.

So how does Spark actually address the shortcomings above? 

By giving you less functionality, Spark ~~encourages~~ forces you to drive your web apps using hypermedia. It is a very thin layer on top of Datastar, which gives you a front-end data store that shares its state with (and is driven by) the back-end. 

Datastar uses [SSE (server-sent events)](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events), which allows a web server to push multiple, real-time updates to the browser over a single HTTP connection. As each template is rendered, it is sent to the web browser via a server-sent event, meaning that elements are streamed in the response and modified in the DOM as soon as the web browser receives them. This may seem like an unusual super-power, but it makes creating highly performant responses possible [needs more explanation].

But what does one do without all the “bells and whistles” that htmx provides? Is this really progress?? Wasn’t “[htmx sucks](https://htmx.org/essays/htmx-sucks/)” only meant as tongue-in-cheek?!?

The idea with Datastar (and Spark) is to go back to first principles while moving the web forward. Web components should be used (natively or using [Lit](https://lit.dev/)) to provide reusable, encapsulated, and framework-agnostic custom elements. Because when you embrace custom elements you are embracing hypermedia.

---

If you are interested in [hypermedia driven apps](https://data-star.dev/guide/howl) (and concepts such as [HOWL](https://htmx.org/essays/hypermedia-on-whatever-youd-like/) and [HATEOS](https://htmx.org/essays/hateoas/), I encourage you to give Datastar and Spark a go. **I’m specifically interested in your thoughts on the Spark API and its approach to performing actions and modifying the DOM and state.**

In my experience, adopting this approach means unlearning a lot of the concepts that using Sprig has spoiled me with. Thanks go to [Delaney Gillilan](https://github.com/delaneyj), author of Datastar, who has patiently answered my questions and addressed my reservations, and who has persistently reminded me that “just because you can, doesn’t mean you should”.

**Comment by delaneyj** [comment]
I think this is amazing and will elevated the usage of PHP forward a ton!

**Reply by bencroker** [reply]
I’ve extracted most of the Datastar specific functionality to a framework agnostic PHP package:
https://github.com/putyourlightson/datastar-php

**Comment by bencroker** [comment]
I’ve created a Spark demo of an inline table editing app. It was a fun process and took me around half an hour to build. It uses signals (via `data-store`) for most of its interactivity, making it extremely snappy to use. Only on saving an entry is a `POST` request made to the server, which responds with an element containing the new title, an element containing an empty alert (to clear out errors), and resetting the `store` value of `entryId` to `0`. If there’s an error then it sends back the first errors in an alert.

This nicely demonstrates the power of signals combined with real-time Twig template rendering, resulting in only the parts of the DOM that need swapping being sent down over the wire.

Go check out the [source code](https://gist.github.com/bencroker/ab131d19d19f841098669bd1b2cdd6ab) (and run it on your own Craft development environment).

https://github.com/user-attachments/assets/c5625934-6a10-4447-94fb-46207a048083

> [!NOTE]
> `editingId` (as seen in the video) was changed to `entryId` in the source code, for clarity.

**Reply by delaneyj** [reply]
This is super exciting!

**Reply by wsydney76** [reply]
Playing around a bit with this example, but struggling with the concept of a 'global store'.

Please correct me if I have misunderstood/overlooked something.

Added `<div data-store="{ slug: 'abc' }">some totally unrelated fancy stuff</div>` somewhere else in the page.

When the server round trip is triggered, all data in the store, including `slug` is sent, causing `craft.app.runAction('entries/save-entry')` to take it into acount and (unintentionally) updating the entries slug.

So what would be the best approach to avoid such conflicts (besides pseudo namespacing data), compared to those other frameworks where each component data (would be each row in this example) is isolated by default?

**Reply by bencroker** [reply]
Please update to 0.0.11 and use [`spark.runAction`](https://putyourlightson.com/plugins/spark#spark-runaction) for running actions.

You’re right that the store values are sent as part of the request body, so the `entries/save-entry` controller action picks them up automatically. I could remove them from the body, but then you’d be required to pass every parameter manually.

Currently:

```twig
{% set response = spark.runAction('users/save-entry') %}
```

With body params removed:

```twig
{% set response = spark.runAction('users/save-entry', {
    entryId: store.entryId,
    title: store.title,
    slug: store.slug,
    ...
}) %}
```

Alternatively, if you want the entire store:

```twig
{% set response = spark.runAction('users/save-entry', store.getValues()) %}
```

**Reply by delaneyj** [reply]
Not sure if this helps but using a _prefix means that signal won't be sent

**Reply by bencroker** [reply]
Good to know, thanks!

**Reply by wsydney76** [reply]
@bencroker Thanks, im using this now, passing only the relevant params:

```twig
{% set response = spark.runAction('entries/save-entry', {
    entryId: store.entryId,
    title: store.title,
}) %}
```

This slug is still updated however, so query params are not completely replaced, but merged?

> Not sure if this helps but using a _prefix means that signal won't be sent

@delaneyj Yes, looks like that would help for frontend-only things.

PS.
Would be a fun thing to make this example real-world-usable in Spark, but doubt I'll find time/skills/energy to work on it. For now I have thrown together a reference Alpine implementation, mostly recycling existing logic/code, that shows how it might look like in the end.

[Gist](https://gist.github.com/wsydney76/60d4be815fd52c3b9be4c2ff5c12ae53)

**Reply by bencroker** [reply]
Yes, because you’re running the same version of Spark. I just released 0.0.12, in which the store is _no longer_ passed through to controller actions, so they _must_ be passed in the second argument of `spark.runAction`.

Quickly looking at your Gist, my guess is that you could end up deleting around half of your code when using Spark.

**Reply by delaneyj** [reply]
Half is a good number!

**Reply by wsydney76** [reply]
@bencroker Thanks, works now in 0.0.12.

>  my guess is that you could end up deleting around half of your code when using Spark.

Hm, not sure where that should come from...

\- drop fetch command
\- drop translation handling in JS
\+ add unsupported stuff like `x-trap.inert.noscroll` (no, not willing to drop that)
\+ whatever it takes to live in the context of a 'global store'
\+/\- all other things like calling the server, evaluating action response, displaying errors, fine grained keyboard and focus control etc. etc. have to be coded somewhere in some way. Suspecting that every x-something becomes a data-something with a comparable amount of code behind it.
\----------------------------------
\= only an equally featured port will tell.

(Having no custom backend component at all in this example is hard to beat...)

PS. 
Either way, you will spend most of your time going through 47 iterations with your client.

**Reply by bencroker** [reply]
> Having no custom backend component at all in this example is hard to beat...

With close to 200 lines of JavaScript code, I tend to disagree.

**Comment by mandrasch** [comment]
Great write-up! 👏👏

Shared this with some other craft devs and one question that came back was: What are possible performance implications with SSE (compared to something like Sprig with regular fetch/ajax calls)? 

Couldn't find much info about it online.

**Reply by bencroker** [reply]
Spark makes a `fetch` request to the server and receives a plaintext response formatted as SSE. Since the response can contain multiple fragments, it means that you only send back what needs to be swapped (rather than an entire component) and you can replace as many parts of the DOM as you like in a single call. So the performance of Spark will generally be better than that of Sprig, although I think you’ll find the gains to be negligible when compared to network latency and database queries.

**Reply by croxton** [reply]
Any thoughts on this? Is it actually relevant to what D* does?

https://dev.to/miketalbot/server-sent-events-are-still-not-production-ready-after-a-decade-a-lesson-for-me-a-warning-for-you-2gie

**Reply by bencroker** [reply]
I’ll need to look into it further, but this is probably something to more relevant to the Datastar Discord server. Note also that this article is from 2020.

**Reply by croxton** [reply]
Yes from the comments it may not really be relevant anymore, but I've never used SSE in production so it's outside my expertise. It may be worth seeking Matt Gray's opinion as an infrastructure expert, since there may be other implications around how SSE interacts with Nginx, static caching and the use of Cloudflare as a proxy across Servd's network (and I guess he'll need to deal with it anyway if Spark gets popular).

**Reply by bencroker** [reply]
I doubt, it – SSE is simpler than you may think! This is what it looks like in a response.
```
event: datastar-signal
data: store {"entryId":0,"alert":true}

event: datastar-fragment
data: fragment
data: <div id="alert" data-show="$alert" class="alert alert-success">
data:         <span>
data:             Entry successfully saved!
data:         </span>
data:     </div>
```

**Reply by mandrasch** [reply]
When I read about Websockets vs SSE this was mentioned: 

> There is also a problem that impacts both technologies - long-running connections.
> 
> Both technologies need to maintain long-lived open connections to send a continuous stream of events.
>
> Managing such connections, especially on a large scale, can be a challenge as we can quite quickly run out of resources. Additionally, they may require special configurations like extended timeout and are more exposed to any network connection problems.

https://softwaremill.com/sse-vs-websockets-comparing-real-time-communication-protocols/

Or does this not apply to datastar/Spark?

**Reply by croxton** [reply]
Helpful answer here on using SSE on Cloudflare and Nginx:

https://community.cloudflare.com/t/using-server-sent-events-sse-with-cloudflare-proxy/656279

**Reply by bencroker** [reply]
Spark renders the template, sends a response and closes the connection, just like a regular response in Craft. Please don't get too caught up on the SSE thing, it's just an implementation detail that Spark handles for you.

**Reply by mandrasch** [reply]
Ah okay, thanks very much for clarification 👍

In CraftQuest Livestream the progress bar example was shown https://data-star.dev/examples/progress_bar, therefore I was interested in this.

And I read that some webspace providers (especially shared hosting) do not allow listening for Websocket connections. So I was unsure about SSE. 🤷

This here might be a special case:  https://www.digitalocean.com/community/questions/does-app-platform-support-sse-server-sent-events-application

A statement from a hosters like "SSE with no long running connections will be fine" would be re-assuring to dive deeper into this approach, at least from my personal perspective. 

And do I understand it correctly that Spark uses SSE mainly because of lightweight protocol, not because of real time updates? Because real time updates are mentioned as primary advantage for SSE in every article. Guess that's what's confusing me. 

Or is it just because datastar uses it and the main advantage of datastar is htmx + alpine in one bundled solution? 🤔

Update: Just saw you added " [needs more explanation]." in the original post, I'll wait for this section to be updated. 

Much further success with it!

**Reply by bencroker** [reply]
Datastar uses SSE for back-end requests, hence SSE. I may explore taking advantage of SSE more in the future, but for now, the benefit is the ability to send multiple fragments, update store values, and use [helper functions](https://putyourlightson.com/plugins/spark#spark-helper-functions) all in the same response.
